src/engine/Union.cpp

```diff
@@ -347,172 +346,250 @@ struct Wrapper {
+      ColumnIndex origin = columnOrigins_.at(column).at(!left);
+      resultTable_.at(resultTable_.size() - 1, column) =
+          origin == Union::NO_COLUMN ? Id::makeUndefined() : row[origin];
     }
-    if (index1 == it1->idTable_.size()) {
-      ++it1;
-      index1 = 0;
+  }
+
+  // ___________________________________________________________________________
+  void advanceRangeIfConsumed() {
+    if (index1_ == it1_.value()->idTable_.size()) {
+      ++it1_.value();
+      index1_ = 0;
     }
-    if (index2 == it2->idTable_.size()) {
-      ++it2;
-      index2 = 0;
+    if (index2_ == it2_.value()->idTable_.size()) {
+      ++it2_.value();
+      index2_ = 0;
     }
   }
 
-  // append the remaining elements
-  for (auto& pair :
-       processRemaining(computePermutation<true>(), it1, range1.end(),
-                        requestLaziness, index1, resultTable, localVocab)) {
-    AD_CORRECTNESS_CHECK(requestLaziness);
-    co_yield pair;
-  }
-  for (auto& pair :
-       processRemaining(computePermutation<false>(), it2, range2.end(),
-                        requestLaziness, index2, resultTable, localVocab)) {
-    AD_CORRECTNESS_CHECK(requestLaziness);
-    co_yield pair;
+  // ___________________________________________________________________________
+  Result::IdTableVocabPair popResult() {
+    auto result = Result::IdTableVocabPair{std::move(resultTable_),
+                                           std::move(localVocab_)};
+    resultTable_ = IdTable{resultTable_.numColumns(), allocator_};
+    resultTable_.reserve(Union::chunkSize);
+    localVocab_ = LocalVocab{};
+    advanceRangeIfConsumed();
+    return result;
   }
-  if (!requestLaziness) {
-    co_yield Result::IdTableVocabPair{std::move(resultTable),
-                                      std::move(localVocab)};
+
+  // ___________________________________________________________________________
+  std::optional<Result::IdTableVocabPair> get() override {
+    if (aggregatedTableReturned_) {
+      return std::nullopt;
+    }
+    if (!it1_.has_value()) {
+      it1_ = range1_.begin();
+    }
+    if (!it2_.has_value()) {
+      it2_ = range2_.begin();
+    }
+    while (it1_ != range1_.end() && it2_ != range2_.end()) {
+      auto& idTable1 = it1_.value()->idTable_;
+      auto& idTable2 = it2_.value()->idTable_;
+      localVocab_.mergeWith(std::span{&it1_.value()->localVocab_, 1});
+      localVocab_.mergeWith(std::span{&it2_.value()->localVocab_, 1});
+      while (index1_ < idTable1.size() && index2_ < idTable2.size()) {
+        if (isSmaller(idTable1.at(index1_), idTable2.at(index2_))) {
+          pushRow(true, idTable1.at(index1_));
+          index1_++;
+        } else {
+          pushRow(false, idTable2.at(index2_));
+          index2_++;
+        }
+        if (requestLaziness_ && resultTable_.size() >= Union::chunkSize) {
+          return popResult();
+        }
+      }
+      advanceRangeIfConsumed();
```

Comment:
You need to call this at the end of each iteration of the outer loop, as well as when prematurely returning.
I think the correct pattern is to put this call in a cleanup struct, but maybe extend our cleanup structs in a way, that this is only performed if we not exit because of an exception.
Otherwise I would prefer the other `advance..` call before and not inside the `popResult`  function.